File: cmd/checkers/sysmon/src/template.rs
/*
 * Copyright 2025 Carver Automation Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// template.rs - Checker template registration

use anyhow::{Context, Result};
use log::{info, warn};

const DEFAULT_TEMPLATE: &str = include_str!("../config/default_template.json");
const CHECKER_KIND: &str = "sysmon";

/// Registers the checker template with the KV service.
/// This writes the default configuration template to templates/checkers/sysmon.json.
/// Safe to call on every startup - templates are factory defaults and can be overwritten.
pub async fn register_template() -> Result<()> {
    // Try to connect to KV service from environment
    let mut kv_client = match kvutil::KvClient::connect_from_env().await {
        Ok(client) => client,
        Err(e) => {
            warn!(
                "Failed to connect to KV service for template registration: {}. \
                 Skipping template registration (this is non-fatal).",
                e
            );
            return Ok(());
        }
    };

    // Validate the template is valid JSON
    if !serde_json::from_str::<serde_json::Value>(DEFAULT_TEMPLATE).is_ok() {
        anyhow::bail!("Embedded default template is not valid JSON");
    }

    // Write template to KV
    let template_key = format!("templates/checkers/{}.json", CHECKER_KIND);
    match kv_client
        .put(&template_key, DEFAULT_TEMPLATE.as_bytes().to_vec())
        .await
    {
        Ok(_) => {
            info!(
                "Successfully registered checker template at {}",
                template_key
            );
            Ok(())
        }
        Err(e) => {
            warn!(
                "Failed to register template at {}: {}. \
                 This is non-fatal, continuing startup.",
                template_key, e
            );
            Ok(())
        }
    }
}

File: pkg/config/bootstrap/template.go
/*
 * Copyright 2025 Carver Automation Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package bootstrap

import (
	"context"
	"errors"
	"fmt"
	"os"
	"strings"
	"time"

	"google.golang.org/grpc"

	"github.com/carverauto/serviceradar/pkg/config"
	"github.com/carverauto/serviceradar/pkg/logger"
	"github.com/carverauto/serviceradar/proto"
)

var (
	errTemplateDataEmpty    = errors.New("template data is empty")
	errCoreAddressNotSet    = errors.New("CORE_ADDRESS environment variable not set")
	errTemplateRegistration = errors.New("failed to register template with core")
)

const templatePublishTimeout = 5 * time.Second

// ServiceWithTemplateRegistration is a convenience function that combines Service() with
// template registration to core for tooling that needs default config templates.
func ServiceWithTemplateRegistration(
	ctx context.Context,
	desc config.ServiceDescriptor,
	cfg interface{},
	templateData []byte,
	opts ServiceOptions,
) (*Result, error) {
	result, err := Service(ctx, desc, cfg, opts)
	if err != nil {
		return nil, err
	}

	if len(templateData) > 0 {
		if err := registerTemplateWithCore(ctx, desc, cfg, templateData, opts); err != nil {
			if opts.Logger != nil {
				opts.Logger.Warn().
					Err(err).
					Str("service", desc.Name).
					Msg("failed to register configuration template; service will continue")
			}
		}
	}

	return result, nil
}

func registerTemplateWithCore(ctx context.Context, desc config.ServiceDescriptor, _ interface{}, templateData []byte, opts ServiceOptions) error {
	if len(templateData) == 0 {
		return errTemplateDataEmpty
	}

	log := opts.Logger
	if log == nil {
		log = logger.NewTestLogger()
	}

	coreAddr := strings.TrimSpace(os.Getenv("CORE_ADDRESS"))
	if coreAddr == "" {
		return errCoreAddressNotSet
	}

	regCtx, cancel := context.WithTimeout(ctx, templatePublishTimeout)
	defer cancel()

	dialOpts, closeProvider, err := BuildCoreDialOptionsFromEnv(regCtx, opts.Role, log)
	if err != nil {
		return fmt.Errorf("%w: %w", errTemplateRegistration, err)
	}

	conn, err := grpc.NewClient(coreAddr, dialOpts...)
	if err != nil {
		closeProvider()
		return fmt.Errorf("%w: %w", errTemplateRegistration, err)
	}
	defer func() {
		if closeErr := conn.Close(); closeErr != nil {
			log.Warn().Err(closeErr).Msg("failed to close core template connection")
		}
	}()
	defer closeProvider()

	client := proto.NewCoreServiceClient(conn)

	resp, err := client.RegisterTemplate(regCtx, &proto.RegisterTemplateRequest{
		ServiceName:  desc.Name,
		TemplateData: templateData,
		Format:       string(desc.Format),
	})
	if err != nil {
		return fmt.Errorf("%w: %w", errTemplateRegistration, err)
	}

	if !resp.GetSuccess() {
		msg := strings.TrimSpace(resp.GetMessage())
		if msg == "" {
			msg = "core rejected template"
		}
		return fmt.Errorf("%w: %s", errTemplateRegistration, msg)
	}

	log.Debug().
		Str("service", desc.Name).
		Msg("registered configuration template with core")

	return nil
}
